Fix a potential hang when starting after a non-clean shutdown

simonferquel opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

This fixes a hang discovered by Docker Desktop users when migrating to v3.0.0: when docker is stopped abruptly - which can happen on Windows Restart with WSL2, restarting could potentially hang.

**- How I did it**

Previous startup sequence used to call "containerStop" on containers that were persisted with a running state but are not alive when restarting (can happen on non-clean shutdown).
This call was made before fixing-up the RunningState of the container, and tricked the daemon to trying to kill a non-existing process and ultimately hang.

The fix is very simple - just add a condition on calling containerStop.

**- How to verify it**

We need to work on a synthetic test for that. I need guidance there.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

Fixes an issue where docker would hang at startup after an abrupt shutdown.


**- A picture of a cute animal (not mandatory but encouraged)**

![image](https://user-images.githubusercontent.com/1677333/102598251-06912f80-411c-11eb-83b2-41bf9a628c9d.png)

simonferquel committed:
Fix a potential hang when starting after a non-clean shutdown

Previous startup sequence used to call "containerStop" on containers that were persisted with a running state but are not alive when restarting (can happen on non-clean shutdown).
This call was made before fixing-up the RunningState of the container, and tricked the daemon to trying to kill a non-existing process and ultimately hang.

The fix is very simple - just add a condition on calling containerStop.

Signed-off-by: Simon Ferquel <simon.ferquel@docker.com>

AkihiroSuda commented on daemon/daemon.go:
```diff
@@ -348,13 +348,14 @@ func (daemon *Daemon) restore() error {
 				log.WithError(err).Error("failed to restore container with containerd")
 				return
 			}
+			log.Debugf("alive: %v", alive)
```
Is this WIP or intentional addition of debug logs?

simonferquel replied:
I found it useful, but could be removed. Anyway, it only shows up with `-D`

cpuguy83 reviewed (changes requested):
Seems ok.

The `if` statement there has gotten pretty gnarly :(

Can you add a test case?

zallarak commented:
Guys can we get this merged in soon? This totally breaks windows docker. In fact, many of our team has to move off windows temporarily to Linux because WSL2 is broken with this ping @craigloewen-msft

cpuguy83 commented:
How is wsl2 broken by this?
This seems like an edge case when docker is not shutdown correctly.
We will certainly fix it, but I do not understand how it can be so detrimental.

Also, can't people downgrade?

cpuguy83 commented:
It seems like `shutdownContainer` should return here but it is not.

The flow for `shutdownContainer` is to try to stop the container gracefully, if that doesn't work (after the specified timeout) it does `kill -9` on it.
If we get `ESRCH` from the kernel (no such process) it returns early.
Otherwise it waits for the notification that the container is shutdown... unfortunately indefinitely because this was really designed around API users being able to cancel the wait, we don't have that during daemon startup.

So it seems like Docker still has the container marked as "Running" because of improper shutdown (fine).
We try to kill the PID recorded in the container state on disk, it *should* return early saying the process doesn't exist... unfortunately the process does exist because the PID was re-used (either due to reboot or even recycling pids after reaching pid max).
Because the process does exist, docker actually ends up killing that pid and then waiting indefinitely for an event from containerd that the container exited which will never come because the container was never running.

That's my theory anyway.

cpuguy83 committed:
Cleanup container shutdown check and add test

Adds a test case for the case where dockerd gets stuck on startup due to
hanging `daemon.shutdownContainer`

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Digging further with @tiborvass, this seems to only happen for containers in 'restarting' state due to some special case handling deep in the call stack for `shutdownContainer`.

People hit this because likely their container is not working correctly to begin with... that said there does not seem to be a way to recover from this outside of manipulating the on-disk state to set `State.Restarting` to `false` (outside of patching dockerd).

I have pushed a change to this PR to make the `if` block cleaner, added some logging, and a test case.

cpuguy83 reviewed (approved):
LGTM

tiborvass commented:
Small correction on https://github.com/moby/moby/pull/41817#issuecomment-748536277, although he is right that this is related to a complicated codepath due to restarting state, the workaround is not to set `Restarting` to `false` in the on disk state, but to set `Running` to `false`.

tiborvass commented on daemon/daemon.go:
```diff
@@ -345,16 +350,24 @@ func (daemon *Daemon) restore() error {
+			logger(c).Debugf("alive: %v", alive)
+			if !alive {
+				// Process will only be nil here if `containerd.Restore` returned an errdefs.NotFoundError
+				// In this case we just want to cleanup a dead container from containerd.
```
Ambiguous `in this case` the case we're in, or the case talked about in the line above? Will amend commit to clarify this.

simonferquel commented on daemon/kill.go:
```diff
@@ -179,7 +179,9 @@ func (daemon *Daemon) Kill(container *containerpkg.Container) error {
 // killPossibleDeadProcess is a wrapper around killSig() suppressing "no such process" error.
 func (daemon *Daemon) killPossiblyDeadProcess(container *containerpkg.Container, sig int) error {
 	err := daemon.killWithSignal(container, sig)
```
Actually killWithSignal hangs if the container has "IsRunning()" returning true while having pid = 0. We should not fall into this state anymore on startup with the fix, but I think it would be nice to make killWithSignal more robust ?

cpuguy83 replied:
Yes, it is a terrible mess.

thaJeztah commented on daemon/daemon.go:
```diff
@@ -363,14 +376,16 @@ func (daemon *Daemon) restore() error {
 			}
 
 			if c.IsRunning() || c.IsPaused() {
+				logger(c).Debug("Syncing container on disk state with real state")
```
nit: perhaps lowercase to be consistent with other logs?

```suggestion
				logger(c).Debug("syncing container state on disk with actual state")
```

thaJeztah commented on daemon/daemon.go:
```diff
@@ -392,13 +407,15 @@ func (daemon *Daemon) restore() error {
 						log.WithError(err).Error("failed to update stopped container state")
 					}
 					c.Unlock()
+					logger(c).Debug("Set stopped state")
```
This should probably be in an `else`, otherwise this is also logged if we failed to update state.

```suggestion
					logger(c).Debug("set stopped state")
```

Also; do we still need this log, or was it mostly used while debugging this issue?

thaJeztah commented on daemon/daemon.go:
```diff
@@ -392,13 +407,15 @@ func (daemon *Daemon) restore() error {
 				}
 
 				if !alive {
+					logger(c).Debug("Setting stopped state")
```
```suggestion
					logger(c).Debug("setting stopped state")
```

thaJeztah commented on daemon/daemon.go:
```diff
@@ -345,16 +350,24 @@ func (daemon *Daemon) restore() error {
+					}
 				}
 			} else if !daemon.configStore.LiveRestoreEnabled {
+				logger(c).Debug("Shutting down container considered alive by containerd")
```
```suggestion
				logger(c).Debug("shutting down container considered alive by containerd")
```

thaJeztah commented on daemon/daemon.go:
```diff
@@ -345,16 +350,24 @@ func (daemon *Daemon) restore() error {
+				// In this case we just want to cleanup a dead container from containerd.
+				// Later on docker's view of the container state will be updated accorrdingly, i.e. SetStopped.
+				if process != nil {
+					logger(c).Debug("Cleaning up dead container process")
```
```suggestion
					logger(c).Debug("cleaning up dead container process")
```

thaJeztah commented on daemon/kill.go:
```diff
@@ -179,7 +179,9 @@ func (daemon *Daemon) Kill(container *containerpkg.Container) error {
 // killPossibleDeadProcess is a wrapper around killSig() suppressing "no such process" error.
 func (daemon *Daemon) killPossiblyDeadProcess(container *containerpkg.Container, sig int) error {
 	err := daemon.killWithSignal(container, sig)
+	logrus.WithError(err).WithField("pid", container.State.Pid).Error("KILLWITHSIGNAL")
 	if errdefs.IsNotFound(err) {
+		logrus.WithError(err).WithField("pid", container.State.Pid).Error("ISNOTFOUND")
```
Looks like these were temporary for debugging? Are they still needed? (I see there's a debug line below already if there's an error; if it's still useful, probably better to combine

tiborvass commented:
Hitting flakey test #40651
